[expo-battery] Add expo-battery unimodule

ninjaachibi opened the pull request:
# Why

We decided to abstract a couple of the battery-related methods away from the `expo-device` unimodule since it made more sense to have them in their own module with their own events/event-emitters. 

The `expo-battery` unimodule exposes a couple methods that allow the developer to query for, as well as monitor battery information. This becomes important for developers who want to consider battery state or power mode when choosing to send background updates, as well as other use cases that require listening to battery information. 

# How

We used the [guide for creating expo unimodules](https://github.com/expo/expo/blob/master/guides/Creating%20Universal%20Modules.md#creating-universal-modules), and then used the Android and iOS documentation to expose native methods related to battery.

We also referenced the [react-native-device-info](https://github.com/react-native-community/react-native-device-info) library for some methods we implemented. 

# Test Plan

We tested the functions on the `sandbox` app under the `/apps` directory, using the following as our `App.js` to test:

Edit: we made a PR #4907 in which you can test this unimodule on NCL

```js
import * as React from 'react';
import { Text, View, StyleSheet, TouchableOpacity } from 'react-native';
import Constants from 'expo-constants';
import * as Battery from 'expo-battery';

export default class App extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      BatteryJSON: {},
    };
  }

  async componentWillMount() {
    await this.pingBattery();
    this._subscribeBatteryLevel();
    this._subscribeBatteryState();
    this._subscribePowerMode();
  }

  componentWillUnmount() {
    this._unsubscribe();
  }

  pingBattery = async () => {
    let BatteryJSON = {};
    try {
      console.log(Battery);
      let batteryLevel = await Battery.getBatteryLevelAsync();
      let powerState = await Battery.getPowerStateAsync();
      let batteryState = await Battery.getBatteryStateAsync();
      let lowPowerMode = await Battery.getLowPowerModeStatusAsync();
      BatteryJSON = { batteryLevel, powerState, batteryState, lowPowerMode };
      this.setState({ BatteryJSON, batteryLevel, powerState, lowPowerMode, batteryState });
      this.forceUpdate();
    } catch (e) {
      console.log('Trouble getting battery info', e);
    }
  };

  _subscribePowerMode = () => {
    this._subscriptionPowerMode = Battery.watchPowerModeChange(({ lowPowerMode }) => {
      this.setState({ lowPowerMode });
      console.log('powerMode changed!', lowPowerMode);
    });
    console.log('subscibed to watch powerMode');
  };
  _subscribeBatteryState = () => {
    this._subscriptionBatteryState = Battery.watchBatteryStateChange(({ batteryState }) => {
      this.setState({ batteryState });
      console.log('batteryState changed!', batteryState);
    });
    console.log('subscibed to watch batterystate');
  };
  _subscribeBatteryLevel = () => {
    this._subscriptionBatteryLevel = Battery.watchBatteryLevelChange(({ batteryLevel }) => {
      this.setState({ batteryLevel });
      console.log('batteryLevel changed!', batteryLevel);
    });
    console.log('subscibed to watch battery level');
  };

  _unsubscribe = () => {
    this._subscriptionPowerMode && this._subscriptionPowerMode.remove();
    this._subscriptionPowerMode = null;
    this._subscriptionBatteryState && this._subscriptionBatteryState.remove();
    this._subscriptionBatteryState = null;
    this._subscriptionBatteryLevel && this._subscriptionBatteryLevel.remove();
    this._subscriptionBatteryLevel = null;
    console.log('unsubscribed');
  };

  render() {
    let { batteryState, batteryLevel, lowPowerMode } = this.state;
    return (
      <View style={styles.container}>
        <View>
          <TouchableOpacity onPress={this.pingBattery}>
            <Text>Ping Battery</Text>
          </TouchableOpacity>
          <Text>{JSON.stringify(this.state.BatteryJSON, null, 2)}</Text>

          <Text>Subscribed values: {`${JSON.stringify(batteryState, null, 2)}, ${batteryLevel}, ${lowPowerMode}`}</Text>


          <TouchableOpacity onPress={this._subscribeBatteryState}>
            <Text>subscribe batteryState</Text>
          </TouchableOpacity>
          <TouchableOpacity onPress={this._subscribeBatteryLevel}>
            <Text>subscribe batteryLevel</Text>
          </TouchableOpacity>
          <TouchableOpacity onPress={this._subscribePowerMode}>
            <Text>subscribe powerMode</Text>
          </TouchableOpacity>
          <TouchableOpacity onPress={this._unsubscribe}>
            <Text>unsubscribe all</Text>
          </TouchableOpacity>

        </View>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    paddingTop: Constants.statusBarHeight,
    backgroundColor: '#ecf0f1',
    padding: 8,
  },
  paragraph: {
    margin: 24,
    fontSize: 18,
    fontWeight: 'bold',
    textAlign: 'center',
  },
});
```

To test the battery event emitters/listeners, we tested using physical Android and Apple devices. 

* Note: a lot of the methods don't seem to work correctly on the iOS simulator since battery information can't be accessed, but they seem to function correctly on Android emulator

vivianzzhu91 committed:
create expo-battery unimodule

Co-authored-by: Robert Luo <1998.robert@gmail.com>
Co-authored-by: Vivian Zhu <tongyu@usc.edu>

Create battery.md

Update battery.md

[js/ts] remove template and add functions

[js/ts] change template into expo-battery

[android] add three battery-related async functions

[ios] Set up pods

[ios] Set up unimodule

[ios] set up unimodule syntax

converted the EX syntax to UM syntax

[ios] update module name

change name from EXBatteryModule->EXBattery

[ts/js] update event emitter and callbacks

[android] create broadcast receiver for event listener, registerReceiver needs update

update yarn.lock

Update ExpoKit.podspec

[ios] install pods

[android] add event emitter in battery module and rename broadcast receiver

[js] split powerStateChange into lowPowerModeChange

[android] add batteryLevel receiver and event emitters

[ios] Add battery methods

[js] make event names camelcase

[ios/js] update listener names

- change powerState -> batteryState since batteryState is more descriptive (enum with values "unknown", "unplugged", "charging", "full")
- update types names to be more descriptive

[android] update getBatteryStateAsync to return enum type

[js/ios] Update isBatteryChargingAsync -> getBatteryStateAsync

[android] replace isCharging with batteryState and update event name

[ts/js] update batteryState functions

[ios/js] fix getBatteryStateAsync

seems like the last commit (3034c2124) didn't work as planned

[android] fix logic on battery state

[docs] add function description

[docs] update README.md for installation guide

[docs] add event listeners and examples

[ts] remove old PowerStateCallback

[ios] fix batteryStateDidChange

previously the function was sending the entire powerState, modified it to jufst send the batteryState in the message body

tsapeta commented on ExpoKit.podspec:
```diff
@@ -26,10 +26,6 @@ Pod::Spec.new do |s|
-    # We want to inhibit warnings in this pod and
-    # CocoaPods doesn't inhibit warnings for transient dependencies
-    # (it's a dependency of FBSDKCoreKit). Update when updating FBSDKCoreKit,
-    # remove once it stops depending on it.
```
this comment needs to be here, probably also an issue with xdl being too old

tsapeta commented on docs/pages/versions/unversioned/sdk/battery.md:
````diff
@@ -0,0 +1,181 @@
+**Examples**
+
+```js
+import Expo from "expo";
````
```suggestion
import Expo from 'expo';
```

tsapeta commented on docs/pages/versions/unversioned/sdk/battery.md:
````diff
@@ -0,0 +1,181 @@
+
+```js
+import Expo from "expo";
+import React from "react";
````
```suggestion
import React from 'react';
```

tsapeta commented on docs/pages/versions/unversioned/sdk/battery.md:
````diff
@@ -0,0 +1,181 @@
+```js
+import Expo from "expo";
+import React from "react";
+import * as Battery from "expo-battery";
````
```suggestion
import * as Battery from 'expo-battery';
```

tsapeta commented on docs/pages/versions/unversioned/sdk/battery.md:
```diff
@@ -0,0 +1,181 @@
+import Expo from "expo";
+import React from "react";
+import * as Battery from "expo-battery";
+import { StyleSheet, Text, View } from "react-native";
```
```suggestion
import { StyleSheet, Text, View } from 'react-native';
```

tsapeta commented on docs/pages/versions/unversioned/sdk/battery.md:
```diff
@@ -0,0 +1,181 @@
+        batteryLevel: result.batteryLevel
+      });
+    });
+    Battery.getBatteryLevelAsync()
```
do we need to ask for the current battery level here? right after starting a watcher?

tsapeta commented on docs/pages/versions/unversioned/sdk/battery.md:
```diff
@@ -0,0 +1,181 @@
+    flex: 1,
+    marginTop: 15,
+    alignItems: "center",
+    justifyContent: "center"
```
```suggestion
    justifyContent: 'center',
```

tsapeta commented on docs/pages/versions/unversioned/sdk/battery.md:
```diff
@@ -0,0 +1,181 @@
+    marginTop: 15,
+    alignItems: "center",
+    justifyContent: "center"
+  }
```
```suggestion
  },
```

tsapeta commented on docs/pages/versions/unversioned/sdk/battery.md:
```diff
@@ -0,0 +1,181 @@
+  container: {
+    flex: 1,
+    marginTop: 15,
+    alignItems: "center",
```
```suggestion
    alignItems: 'center',
```

tsapeta commented on ios/Podfile.lock:
```diff
@@ -1181,7 +1183,7 @@ DEPENDENCIES:
   - AppAuth (~> 0.4)
   - Bolts (= 1.9.0)
-  - Branch (~> 0.27.1)
+  - Branch (~> 0.24.2)
```
We don't want to downgrade Branch, probably your linked `@expo/xdl` is too old or sth 🤔 Try to run `et ios-generate-dynamic-macros` and then run `pod install` in `ios` folder again.

vivianzzhu91 committed:
[docs] fix indentation and double quote to single quote

vivianzzhu91 committed:
[android] fix indentation in gradle

vivianzzhu91 committed:
[ts/js] remove importing errors and format code styles

ninjaachibi commented on packages/expo-battery/ios/EXBattery/EXBattery.m:
```diff
@@ -0,0 +1,183 @@
+  NSArray *batteryStates = @[@"unknown", @"unplugged", @"charging", @"full"];
+  __block NSDictionary *powerState;
+  
+  [UMUtilities performSynchronouslyOnMainThread:^{
```
Yes, you're right!

ninjaachibi committed:
Fix Battery.js.map

Circle CI was failing because of this, so rebuiltl

ninjaachibi committed:
[ios] Reupgrade branch from 0.24.2->0.27.1

seemed like Branch got downgraded somehow, so reupgrading it to what it should be after running `et ios-generate-dynamic-macros` and then run `pod install` in ios folder again.

ninjaachibi committed:
Merge branch 'master' into @expo-battery

ninjaachibi committed:
Update ExpoKit.podspec

fix ExpoKit warning

ninjaachibi committed:
[ios] remove unnecessary expressions

ninjaachibi committed:
[ios] remove synchronous block

since we implement the methodQueue, tasks should be called on the main queue anyway

ninjaachibi committed:
[ios] update pods

ninjaachibi committed:
Update yarn.lock

ninjaachibi committed:
[ios] workaround for unimodule lifecycle bug

See #4847

ninjaachibi committed:
Add getLowPowerModeStatusAsync()

fix some names and add new method

ninjaachibi committed:
[ios/js/android] Add getLowPowerModeStatusAsync() method

* change lowPowerMode to a return string in android
* add getLowPowerModeStatusAsync() method to all platforms

ninjaachibi committed:
[android] Fix unresolved promise in getLowPowerModeStatusAsync()

ninjaachibi committed:
[ios/js] make event listener callbacks consistent

In the docs, the argument of the callback is an object with a key, so I have made those consistent on the js and iOS sides.

ninjaachibi committed:
Update battery.md

* fix some flow
* add to the EventSubscription examples

ninjaachibi committed:
[android] update batteryState enum to be consistent

the documentation calls for all teh enum values to be lowercase

vivianzzhu91 committed:
[docs] rename battery saver to power saver on android

ninjaachibi committed:
Merge branch 'master' into @expo-battery

ninjaachibi committed:
Update battery.md

more clearly display warning for ios simulator

esamelson commented on packages/expo-battery/android/src/main/java/expo/modules/battery/BatteryLevelReceiver.java:
```diff
@@ -0,0 +1,24 @@
+import android.os.PowerManager;
+
+public class BatteryLevelReceiver extends BroadcastReceiver {
+  public static String INTENT_ACTION = "expo.modules.battery.BatteryLevelReceiver";
```
is this used?

vivianzzhu91 replied:
Yes. This is used for event listener in the battery module. This is triggered when battery level dropped below `"android.intent.action.BATTERY_LOW"` or up to `"android.intent.action.BATTERY_OKAY"` from low battery level.  I just used this `Intent.ACTION_BATTERY_CHANGED` intent filter to get the current battery level and return the value back.

esamelson commented on docs/pages/versions/unversioned/sdk/battery.md:
```diff
@@ -0,0 +1,197 @@
+
+### `Battery.getBatteryLevelAsync()`
+
+Get the battery level of the device as a float between 0 and 1.
```
also, what happens if the battery level cannot be retrieved? On Android, it looks like we pass in -1 as a default value -- should document this.

esamelson commented on packages/expo-battery/android/src/main/java/expo/modules/battery/BatteryModule.java:
```diff
@@ -0,0 +1,173 @@
+      promise.resolve(batteryLevel);
+    } catch (NullPointerException e) {
+      Log.e(TAG, e.getMessage());
+      promise.reject("ERR_BATTERY", "Null Exception in trying to get battery level");
```
```suggestion
      promise.reject("ERR_BATTERY", "Could not get battery level", e);
```
and same in all other methods

esamelson commented on packages/expo-battery/android/src/main/java/expo/modules/battery/BatteryModule.java:
```diff
@@ -0,0 +1,173 @@
+      IntentFilter ifilter = new IntentFilter(Intent.ACTION_BATTERY_CHANGED);
+      Intent batteryStatus = this.mContext.getApplicationContext().registerReceiver(null, ifilter);
+      int status = batteryStatus.getIntExtra(BatteryManager.EXTRA_STATUS, -1);
+      if (status == BatteryManager.BATTERY_STATUS_FULL) {
```
better to extract all of this logic to a separate method, maybe something like `public static BatteryState batteryStatusNativeToJS(int status)` (public so that `BatteryStateReceiver` can use it too)

esamelson commented on packages/expo-battery/ios/EXBattery/EXBattery.m:
```diff
@@ -0,0 +1,191 @@
+//    [[NSNotificationCenter defaultCenter] addObserver:self
+//                                             selector:@selector(powerModeDidChange:)
+//                                                 name:NSProcessInfoPowerStateDidChangeNotification
+//                                               object:nil];
```
remove this commented code

esamelson commented on packages/expo-battery/ios/EXBattery/EXBattery.m:
```diff
@@ -0,0 +1,191 @@
+}
+
+- (void)dealloc {
+//  [self invalidate];
```
should this be uncommented?

ninjaachibi replied:
Yep, I uncommented it.

esamelson commented on packages/expo-battery/ios/EXBattery/EXBattery.m:
```diff
@@ -0,0 +1,191 @@
+  _eventEmitter = [moduleRegistry getModuleImplementingProtocol:@protocol(UMEventEmitterService)];
+  
+  if (moduleRegistry) {
+    [UIDevice.currentDevice setBatteryMonitoringEnabled:YES];
```
are there ramifications of setting this to YES that we should document?

ninjaachibi replied:
I didn't find any specific side effects or ramifications. The official docs state that 
```
Enable battery monitoring if your app needs to be notified of changes to the battery state, or if you want to check the battery charge level.

The default value of this property is NO, which:

Disables the posting of battery-related notifications

Disables the ability to read battery charge level and battery state
```
But I do see in our implementation that when in the uni-module lifecycle, when our `expo-battery` is registered in `setModuleRegistry`, event listeners are also registered for `NSNotificationsCenter`. Is this the desired behavior? From what I understand, perhaps we would only want to register these event listeners when the user chooses to subscribe?

esamelson replied:
cool, thanks for researching this thoroughly. Yes, I think you're right - we probably want to add these observers in the `startObserving` method and then remove them in `stopObserving`. I believe RN takes care of calling those methods when the first listener is added and when the last listener is removed in JS.

esamelson commented on packages/expo-battery/ios/EXBattery/EXBattery.m:
```diff
@@ -0,0 +1,191 @@
+                    getBatteryLevelAsyncWithResolver:(UMPromiseResolveBlock)resolve
+                    rejecter:(UMPromiseRejectBlock)reject)
+{
+  resolve(@([[self getPowerState][@"batteryLevel"] floatValue]));
```
```suggestion
  resolve([self getPowerState][@"batteryLevel"]);
```

esamelson reviewed (changes requested):
sorry for the slow review here, guys 😞 in addition to the comments, could you add a ToC and error codes section to the docs, and add a web implementation as well?

vivianzzhu91 committed:
[docs] add error codes and enums

vivianzzhu91 committed:
[js] rename and add enums

ninjaachibi committed:
Merge branch 'master' into @expo-battery

ninjaachibi committed:
[ios] update pods

ran pod install

vivianzzhu91 committed:
[docs] update exmaple and description

vivianzzhu91 committed:
[android] android rename and enums

vivianzzhu91 committed:
[js] update enums

vivianzzhu91 committed:
[docs] update batteryLevelListener description

ninjaachibi committed:
[ts/js] change isLowPowerModeEnableAsync -> isLowPowerModeEnabledAsync

vivianzzhu91 committed:
restore deference object in event emitters

ninjaachibi committed:
fix isLowPowerModeEnabledAsync()

ninjaachibi committed:
Update yarn.lock

ninjaachibi committed:
[ios] update types and implementations

* lowPowerMode is now a boolean
* capitalized enum values for batteryStates
* removed commented code for unimodule life cycle

vivianzzhu91 committed:
rename isLowPowerModeEnableAsync -> isLowPowerModeEnabledAsync

ninjaachibi committed:
[ios/docs] remove warnings

* [docs] isLowPowerModeEnableAsync -> isLowPowerModeEnabledAsync
* remove warnings for iOS battery monitoring

vivianzzhu91 committed:
update batteryState enum with numbers

esamelson reviewed (commented):
this looks good to me after this last comment is resolved! 👍

ninjaachibi committed:
Merge branch 'master' into @expo-battery

ninjaachibi committed:
[ios] update pods

ninjaachibi committed:
[ios] fix event emitters

put event emitter registration in startObserving() and same thing for unregistrationin stopObserving()

ninjaachibi committed:
Merge branch 'master' into @expo-battery

ninjaachibi committed:
[ios] update pods
